File: services/ai-service/internal_auth.py
"""Internal-mesh authentication for ai-service's gRPC surface.

Every RPC here acts on a caller-supplied ``user_id`` (classify-and-persist, read a user's
recommendations, read/extend a chat thread), so the port must only ever be reachable by the
gateway and other mesh peers. It previously accepted anonymous calls from anyone who could reach
it, which meant anyone could act as any user and spend the project's Gemini quota.

``grpc.health.v1.Health`` stays exempt so the gateway's health checker can probe it.
"""

from __future__ import annotations

import hmac

import grpc

_HEALTH_PREFIX = "/grpc.health.v1.Health/"
_TOKEN_KEY = "x-internal-token"


class InternalAuthInterceptor(grpc.aio.ServerInterceptor):
    def __init__(self, token: str):
        self._token = token

    async def intercept_service(self, continuation, handler_call_details):
        method = handler_call_details.method or ""
        if method.startswith(_HEALTH_PREFIX):
            return await continuation(handler_call_details)

        # Resolved before the auth check (this only looks up the registered handler, it
        # doesn't invoke it) purely so a rejection can be shaped to match the RPC's real
        # cardinality — ChatStream is unary_stream, and aborting it with a unary_unary
        # handler would desync the client's stream reader instead of surfacing a clean
        # UNAUTHENTICATED.
        handler = await continuation(handler_call_details)

        if not self._token:
            return _abort_handler(
                handler,
                grpc.StatusCode.FAILED_PRECONDITION,
                "Internal service token is not configured.",
            )

        metadata = dict(handler_call_details.invocation_metadata or ())
        presented = metadata.get(_TOKEN_KEY, "")

        # hmac.compare_digest is constant-time, so this can't be turned into a byte-at-a-time
        # oracle for the token.
        if not hmac.compare_digest(str(presented), self._token):
            return _abort_handler(
                handler,
                grpc.StatusCode.UNAUTHENTICATED,
                "This endpoint is restricted to internal mesh callers.",
            )

        return handler


def _abort_handler(handler, code: grpc.StatusCode, details: str):
    if handler is not None and handler.response_streaming:
        async def abort_stream(request, context):
            await context.abort(code, details)
            return
            yield  # pragma: no cover - makes this an async generator, never reached

        return grpc.unary_stream_rpc_method_handler(abort_stream)

    async def abort_unary(request, context):
        await context.abort(code, details)

    return grpc.unary_unary_rpc_method_handler(abort_unary)

File: services/messaging-service/src/grpc/internalAuth.js
const crypto = require('crypto');
const grpc = require('@grpc/grpc-js');
const env = require('../config/env');

// Requires the mesh's shared internal token on an inbound gRPC call, and returns the end user
// the internal caller (the gateway) is acting for.
//
// This port previously accepted calls from anyone who could reach it, with no authentication of
// any kind, while all end-user authentication lived in the gateway. Trusting x-user-id is only
// sound *after* the token check has established that the caller is the gateway, which validated
// the user's JWT itself.
function requireInternalCaller(call) {
  if (!env.internalServiceToken) {
    throw { code: grpc.status.FAILED_PRECONDITION, details: 'Internal service token is not configured.' };
  }

  const presented = String(call.metadata.get('x-internal-token')[0] || '');
  if (!fixedTimeEquals(presented, env.internalServiceToken)) {
    throw { code: grpc.status.UNAUTHENTICATED, details: 'This endpoint is restricted to internal mesh callers.' };
  }

  const userId = String(call.metadata.get('x-user-id')[0] || '');
  if (!userId) {
    throw {
      code: grpc.status.UNAUTHENTICATED,
      details: "Calls must carry the acting user's id in x-user-id metadata.",
    };
  }

  return userId;
}

function fixedTimeEquals(a, b) {
  const bufA = Buffer.from(a, 'utf8');
  const bufB = Buffer.from(b, 'utf8');
  if (bufA.length !== bufB.length) return false;
  return crypto.timingSafeEqual(bufA, bufB);
}

module.exports = { requireInternalCaller };
